proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+#### Tag Types
+
+```
+tagtype ::= [t*]→[]
+```
+
+### Instructions
+
+```
+instr ::= ... | throw tagidx | rethrow labelidx
+        | try blocktype instr* (catch tagidx instr*)* (catch_all instr*)? end
+        | try blocktype instr* delegate labelidx
+```
+
+### Modules
+
+#### Tags
+
+```
+tag ::= export* tag tagtype  | export* tag tagtype import
+```
+
+#### Modules
+
+```
+mod ::= module ... tag*
+```
+
+## Validation (Typing)
+
+#### Modification to labels
+
+To verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we introduce a `kind` attribute to labels in the validation context, which is set to `catch` when the label is a catch-label and empty otherwise.
+
+```
+labelkind ::= catch
+labeltype ::= {result resulttype, kind labelkind?}
+C ::= {..., labels labeltype}
+```
+
+The original notation `labels [t*]` is now an abbreviation for:
+
+```
+labels [t*] ::= labels {result [t*], kind ε}
+```
+
+
+### Instructions
+
+
+```
+C.tags[x] = [t*]→[]
+-----------------------------
+C ⊢ throw x : [t1* t*]→[t2*]
+
+
+C.labels[l].kind = catch
+----------------------------
+C ⊢ rethrow l : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+(C.tags[x] = [t*]→[] ∧
+ C, labels { result [t2*], kind catch } ⊢ instr'* : [t*]→[t2*])*
+(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+C.labels[l] = [t*]
+-------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+#### Stores
+
+```
+S ::= {..., tags taginst*}
+```
+
+#### Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+#### Module Instances
+
+```
+m ::= {..., tags tagaddr*}
+```
+
+#### Administrative Instructions
+
+```
+instr ::= ... | throw tagaddr | catch_n{ tagaddr? instr* }* instr* end
+        | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end
+```
+
+#### Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts.
+
+```
+B^0 ::= val* [_] instr*
+B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end
+      | delegate{ labelidx } B^k end
+B^{k+1} ::= val* (label_n{instr*} B^k end) instr*
```

Comment:
Thanks for the suggestion on the block contexts, I will add it in the next update. 

> Btw, is there a reason that catch and caught have an arity, but delegate does not?

I can't remember any more tbh, although it might be necessary to reduce a `delegate{l} val^n end`. Will add it in the next update as well.

Reply:
> I can't remember any more tbh, although it might be necessary to reduce a `delegate{l} val^n end`. Will add it in the next update as well.

The `n` there doesn't matter, you could just use `*`. So I don't think it's needed. I am almost sure that `catch` and `caught` do not need it either.

(The reason it's needed for `label` and `frame` is branches and return instructions: the rules for those need to know how many operands to transport along.)

Reply:
Thank you for the clarification! That's good, I will remove all of them then. By the way, do you think that using the same names `delegate` and `catch` for the administrative instructions can get confusing? 
What do you think about using `delegation` and `handler' respectively, instead? (Or any other names, these are just examples).

Reply:
It may be somewhat confusing, but I suspect that inventing different names might just obscure the role of the instructions even more.

FWIW, this question also comes up in the threads proposal, which has more administrative instr's that internalise corresponding source instr's. Would be nice to have a generic naming scheme for this that makes the relation immediately clear. (So far I used `name'` for internal counterparts, but the tick is a bit too low a signal.)

Reply:
I think the name sharing is fine; coming up with a different set of names just to be different can be awkward and harder to memorize.